src/main/java/com/limechain/babe/BlockProductionVerifier.java

```diff
@@ -171,4 +188,33 @@ private Authority getAuthority(List<Authority> authorities, int index) {
         }
         return authorities.get(index);
     }
+
+    private boolean isBlockEquivocationExist(byte[] authorityPublicKey,
+                                           BlockHeader blockHeader,
+                                           Runtime runtime,
+                                           BigInteger currentSlotNumber) {
+        String hexPublicKey = HexUtils.toHexString(authorityPublicKey);
+        if (currentSlotAuthorBlockMap.containsKey(hexPublicKey)) {
+            BlockHeader firstBlockHeader = currentSlotAuthorBlockMap.get(hexPublicKey);
+            if (!firstBlockHeader.equals(blockHeader)) {
+                BlockEquivocationProof blockEquivocationProof = new BlockEquivocationProof();
+                blockEquivocationProof.setPublicKey(authorityPublicKey);
+                blockEquivocationProof.setSlotNumber(currentSlotNumber);
+                blockEquivocationProof.setFirstBlockHeader(firstBlockHeader);
+                blockEquivocationProof.setSecondBlockHeader(blockHeader);
+
+                Optional<OpaqueKeyOwnershipProof> opaqueKeyOwnershipProof = runtime.generateKeyOwnershipProof(currentSlotNumber, authorityPublicKey);
+                opaqueKeyOwnershipProof.ifPresentOrElse(
+                        key -> runtime.submitReportEquivocationUnsignedExtrinsic(blockEquivocationProof, key.getProof()),
+                        () -> log.warning(String.format("Unable to generate Opaque Key Ownership Proof for authority: %s", hexPublicKey)));
```

Comment:
```suggestion
                        () -> log.warning(String.format("Failure to report equivocation for authority: %s. Authorship verification marked as failure.", hexPublicKey)));
```

I feel like we should note that authorship verification fails even if the equivocation doesn't get reported.